src/main/java/org/jabref/logic/bst/util/BstTextPrefixer.java

```diff
@@ -23,49 +23,65 @@ public class BstTextPrefixer {
-        int n = cs.length;
-        int i = 0;
-
-        int braceLevel = 0;
+        while (prefixState.index < cs.length && prefixState.numOfChars > 0) {
+            char c = cs[prefixState.index++];
+            handleOpeningBrace(cs, prefixState, c);
+            handleClosingBrace(prefixState, toPrefix, c);
+            boolean isNormalCharacter = c != '{' && c != '}';
+            if (isNormalCharacter) {
+                prefixState.numOfChars--;
+            }
+        }
+        sb.append(toPrefix, 0, prefixState.index);
+        // Append any remaining closing braces if unbalanced
+        while (prefixState.braceLevel-- > 0) {
+            sb.append('}');
+        }
+        return sb.toString();
+    }
 
-        while ((i < n) && (numOfChars > 0)) {
-            char c = cs[i];
-            i++;
-            if (c == '{') {
-                braceLevel++;
-                if ((braceLevel == 1) && (i < n) && (cs[i] == '\\')) {
-                    i++; // skip backslash
-                    while ((i < n) && (braceLevel > 0)) {
-                        if (cs[i] == '}') {
-                            braceLevel--;
-                        } else if (cs[i] == '{') {
-                            braceLevel++;
-                        }
-                        i++;
-                    }
-                    numOfChars--;
-                }
-            } else if (c == '}') {
-                if (braceLevel > 0) {
-                    braceLevel--;
-                } else {
-                    LOGGER.warn("Unbalanced brace in string for purify$: {}", toPrefix);
+    private static void handleOpeningBrace(char[] cs, PrefixState prefixState, char c) {
+        if (c != '{') {
+            return;
+        }
+        prefixState.braceLevel++;
+        if ((prefixState.braceLevel == 1) && (prefixState.index < cs.length) && (cs[prefixState.index] == '\\')) {
+            prefixState.index++; // skip backslash
+            while ((prefixState.index < cs.length) && (prefixState.braceLevel > 0)) {
+                if (cs[prefixState.index] == '}') {
+                    prefixState.braceLevel--;
+                } else if (cs[prefixState.index] == '{') {
+                    prefixState.braceLevel++;
                 }
-            } else {
-                numOfChars--;
+                prefixState.index++;
             }
+            prefixState.numOfChars--;
         }
-        sb.append(toPrefix, 0, i);
-        while (braceLevel > 0) {
-            sb.append('}');
-            braceLevel--;
+    }
+
+    private static void handleClosingBrace(PrefixState prefixState, String toPrefix, char c) {
+        if (c != '}') {
+            return;
         }
+        if (prefixState.braceLevel > 0) {
+            prefixState.braceLevel--;
+        } else {
+            LOGGER.warn("Unbalanced brace in string for purify$: {}", toPrefix);
+        }
+    }
 
-        return sb.toString();
+    private static class PrefixState {
```

Comment:
For such cases it's better to use a record class that is much easier

Reply:
Great idea! I'll try that later.

Reply:
> For such cases it's better to use a record class that is much easier

I checked the `record` example, find records are immutable by design. It may not suitable in this code because I need to change `index`, `braceLevel` and `numOfChars` value.